src/udx.c

```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
   // TODO: retransmits are counted in pkts_waiting, but we (prob) should not count them
   // towards to drain loop - investigate that.
   if (was_waiting > 0 && stream->pkts_waiting == 0 && stream->on_drain != NULL) {
+    // todo: move this callback into the fill window call?
     stream->on_drain(stream);
   }
 
   if (sent < 0) return sent;
   return 0;
 }
 
+static inline uint32_t
+get_window_bytes (udx_stream_t *stream) {
+  // todo: receive window, then return lesser of cwnd, rwnd
+  // todo:
+  if (stream->inflight >= stream->cwnd * UDX_MSS) {
+    return 0;
+  }
+  return stream->cwnd * UDX_MSS - stream->inflight;
+}
+
+static int
+fill_window (udx_stream_t *stream) {
+
+  if (stream->pkts_waiting > 0) {
+    int rc = flush_waiting_packets(stream);
+    if (rc < 0) {
+      return rc;
+    }
+    // if packets are still waiting, no need to drain buffers
+    if (stream->pkts_waiting > 0) {
+      return 0;
+    }
+  }
+
+  while (get_window_bytes(stream) > 0 && stream->write_buffer_queue.len > 0) {
+    udx_write_buffer_t *wbuf = udx__fifo_first(&stream->write_buffer_queue);
+    assert(wbuf != NULL);
+    uv_buf_t *buf = &wbuf->buf;
+    assert(buf != NULL);
+
+    udx_stream_write_t *w = wbuf->write;
+
+    int header_flag = wbuf->is_write_end ? UDX_HEADER_END : UDX_HEADER_DATA;
+
+    int len = get_window_bytes(stream);
+    if (buf->len < len) len = buf->len;
+    if (stream->mtu < len) len = stream->mtu;
+
+    if (stream->mtu > len && buf->len > len && stream->pkts_inflight > 0) {
```

Comment:
when is this true given the clamping above?

Reply:
in the last iteration of the loop, if neither of the clamps are triggered, e.g. remaining window size is less than the remaining buffer, and remaining window size is less than the stream mtu value.

the conditional there bails out of sending the small packet at the end of the flight, unless there are zero packets in flight and we need to send a small packet to make progress (e.g. the whole write is only one packet). It's an optimization kind of like TCP Cork.

Reply:
Does this delay sending a packet? Cause we rely on packets not being queued unless they have to (ie no naggle algo)